openam-forgerock-rest/src/main/java/org/forgerock/openam/forgerockrest/IdentityResource.java

```diff
@@ -644,21 +645,19 @@ private void anonymousUpdate(final ServerContext context, final ActionRequest re
         String tokenID;
         String confirmationId;
         String username;
-        String nwpassword;
+        String newPassword;
         final JsonValue jVal = request.getContent();
 
         try{
             tokenID = jVal.get(TOKEN_ID).asString();
-            jVal.remove(TOKEN_ID);
             confirmationId = jVal.get(CONFIRMATION_ID).asString();
-            jVal.remove(CONFIRMATION_ID);
             username = jVal.get(USERNAME).asString();
-            nwpassword =  jVal.get("userpassword").asString();
+            newPassword =  jVal.get(USER_PASSWORD).asString();
 
             if(username == null || username.isEmpty()){
                 throw new BadRequestException("username not provided");
             }
-            if(nwpassword == null || username.isEmpty()) {
+            if (newPassword == null || newPassword.isEmpty()) {
                 throw new BadRequestException("new password not provided");
             }
 
```

Comment:
This fix is not complete, https://github.com/aldaris/OpenAM/commit/d08466f8cb67b699ebcfb6052b348f7b7fc5c00b#diff-343b84b79e4cbc68ed84eccc7a80fc1fR830 plays key role in the fix.

Reply:
Is it the "json(object(field(USERNAME, username), field(USER_PASSWORD, newPassword))" that's important or the "thenAsync"?

Reply:
former